pkg/client/daemon/tunrouter.go

```diff
@@ -92,37 +102,94 @@ type tunRouter struct {
 	// session contains the manager session
 	session *manager.SessionInfo
 
-	// mgrConfigured will be closed as soon as the connector has sent over the correct port to
+	// cfgComplete will be closed as soon as the connector has sent over the correct port to
 	// the traffic manager and the managerClient has been connected.
-	mgrConfigured <-chan struct{}
+	cfgComplete chan struct{}
 
 	// rndSource is the source for the random number generator in the TCP handlers
 	rndSource rand.Source
 }
 
-func newTunRouter(managerConfigured <-chan struct{}) (*tunRouter, error) {
+func newTunRouter() (*tunRouter, error) {
 	td, err := tun.OpenTun()
 	if err != nil {
 		return nil, err
 	}
 	return &tunRouter{
-		dev:           td,
-		handlers:      connpool.NewPool(),
-		toTunCh:       make(chan ip.Packet, 100),
-		mgrConfigured: managerConfigured,
-		fragmentMap:   make(map[uint16][]*buffer.Data),
-		rndSource:     rand.NewSource(time.Now().UnixNano()),
+		dev:         td,
+		handlers:    connpool.NewPool(),
+		toTunCh:     make(chan ip.Packet, 100),
+		cfgComplete: make(chan struct{}),
+		fragmentMap: make(map[uint16][]*buffer.Data),
+		rndSource:   rand.NewSource(time.Now().UnixNano()),
 	}, nil
 }
 
+func (t *tunRouter) configured() <-chan struct{} {
+	return t.cfgComplete
+}
+
 func (t *tunRouter) configureDNS(_ context.Context, dnsIP net.IP, dnsPort uint16, dnsLocalAddr *net.UDPAddr) error {
 	t.dnsIP = dnsIP
 	t.dnsPort = dnsPort
 	t.dnsLocalAddr = dnsLocalAddr
 	return nil
 }
 
-func (t *tunRouter) setOutboundInfo(ctx context.Context, mi *daemon.OutboundInfo) (err error) {
+func (t *tunRouter) refreshSubnets(ctx context.Context) error {
+	desiredSubnets := make([]*net.IPNet, len(t.clusterSubnets)+len(t.alsoProxySubnets))
+	copy(desiredSubnets, t.clusterSubnets)
+	copy(desiredSubnets[len(t.clusterSubnets):], t.alsoProxySubnets)
+	desiredSubnets = subnet.Unique(desiredSubnets)
+
+	// Trim those that are already routed
+	ln := len(desiredSubnets)
+	for i, desiredSn := range desiredSubnets {
+		if i >= ln {
+			break
+		}
+		for _, routedSn := range t.subnets {
+			if subnet.Equal(desiredSn, routedSn) {
+				ln--
+				desiredSubnets[i] = desiredSubnets[ln]
+				break
+			}
+		}
+	}
+	desiredSubnets = desiredSubnets[:ln]
```

Comment:
I'm not asking you to change it, and maybe it's just me, but I found this code hard to read.  I'm not sure what I'd do about that without making it less efficient. :shrug:

Reply:
I guess maybe I'm a little confused so maybe adding a comment would be helpful...

but could you check my understanding of the inner for loop.  If the desiredSn (at index i), is already in t.subnets, then we want to "trim" it.  So we decrement the length of the desired subnets by 1 since we are going to "remove" one.  

But in  `desiredSubnets[i] = desiredSubnets[ln]`, we set the ith index to be the new decremented lnth element, but since that element is now in position `i`, will we never check if desiredSubnets[ln] is already included right?

Reply:
So "ln" is short for "length". When something that should be removed is discovered at position `i`, then that element is overwritten with the last element in the list (effectively `ln - 1` , but `ln` is decreased on the preceding line instead of after). The decrease of `ln` is necessary because the last element is now a duplicate (it's also at position `i`).

At the end of the loop, the slice is truncated using `[:ln]` which effectively removes all the duplicates.

Reply:
I'll try and come up with comments that explains what goes on.